Fix module loading to ensure each module is loaded only once

pgammans opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
This PR addresses the issues found in the issues #22842, PR #26557,  PR #36760, PR #26557

It also replaces the PR for #26557 which has a number of issues and is an 


By forcing the subscription to close if a module has a lazy sub-module then the these are passed to the PreloadingStrategy class as the observable has already terminated.
It doesn't actually prevent the load as the fetch may overlap with the preload fetch. Ie both the pre-loader and navigation can start a load both will finish.
Issue Number: #26557, #22842


## What is the new behavior?
By Prevent loading a route module twice, we prevent the sending of multiple RouteConfigLoadEnd events, and the duplicate console log as show in the stackbiz example on #26557

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
This is an updated version of my originally and closed PR #36760 (due to my failure to update) but addressing @atscott comments.

pgammans committed:
fix(router): fix load interaction of navigation and preload strategies

Fix router to ensure that a route module is only loaded once especially
in relation to the use of preload strategies with delayed or partial
loading.

Add test to check the interaction of PreloadingStrategy and normal
router navigation under differing scenarios.
Checking:
 * Prevention of duplicate loading of modules.
   related to #26557
 * Prevention of duplicate RouteConfigLoad(Start|End) events
   related to #22842
 * Ensuring preload strategy remains active for submodules if needed
   The selected preload strategy should still decide when to load submodules
 * Possibility of memory leak with unfinished preload subscription
   related to #26557
 * Ensure that the stored loader promise is cleared so that subsequent
   load will try the fetch again.
 * Add error handle error from loadChildren
 * Ensure we handle error from with NgModule create

Fixes #26557 #22842 #26557

atscott commented on packages/router/src/router_config_loader.ts:
```diff
@@ -28,22 +28,29 @@ export class RouterConfigLoader {
+          return new LoadedRouterConfig(
+              flatten(module.injector.get(ROUTES)).map(standardizeConfig), module);
+        }),
+        share());
```
I'm wondering if there's a way we can avoid adding the `share` operator to this fix. The router doesn't currently use it so by adding it here, we'll be increasing the bundle size by a decent bit since `share` is a non-trivial operator.

Maybe we can do something like this instead
```
    const loadedConfigSubject = new Subject<LoadedRouterConfig>();
    route._loader$ = loadedConfigSubject.asObservable();
    return moduleFactory$.pipe(
        catchError((err) => {
          loadedConfigSubject.error(err);
          loadedConfigSubject.complete();
          route._loader$ = undefined;
          throw err;
        }),
        map((factory: NgModuleFactory<any>) => {
         ...
        }),
        tap(loadedConfig => {
          loadedConfigSubject.next(loadedConfig);
          loadedConfigSubject.complete();
          route._loader$ = undefined;
        }));
```
WDYT?

pgammans replied:
I have implement this and it passes the tests, but in doing so have realise there is a small change in behaviour.

If there is a exception in the map function in the code above,  ie this could be in either the 
 - onLoadEndListener
 - factory creation of the module
 - creating the new LoadedRouterConfig

we reuse the same loaded observable as _loader is not cleared.
I have written a test to catch this and we can simply be fix it by moving the catchError to the end of the moduleFactory$.pipe. ie  map, tap, catchError

But am unsure what you think the correct behaviour of this code should be
ie 
1. If the factory fails should we still need to call onLoadEndListener thus sending 'RouteConfigLoadEnd' (master does for error in the map but not for error on the loader)
2. If the factory fails should we reload the module?


I commit code to keep the current behaviour be this we should only call onLoadEndListener if success from the factory and building the LoadedRouterConfig.

atscott replied:
I think it's pretty unlikely that a successfully retrieved factory would error on creation. But I think what you've done by moving the `catchError` to the end so that reloading is attempted on the next call is still a good idea just in case 👍

atscott commented on packages/router/test/router_preloader.spec.ts:
```diff
@@ -2,13 +2,17 @@
  * @license
  * Copyright Google LLC All Rights Reserved.
  *
+
```
nit: looks like this change was a mistake

pgammans replied:
yes sorry.

atscott commented on packages/router/test/router_preloader.spec.ts:
```diff
@@ -196,6 +200,272 @@ describe('RouterPreloader', () => {
+
+    beforeEach(() => {
+      delayLoadUnPasued = new BehaviorSubject<string[]>([]);
+      delayLoadObserver$ = delayLoadUnPasued.asObservable();
```
```suggestion
      delayLoadObserver$ = delayLoadUnPaused.asObservable();
```

atscott commented on packages/router/test/router_preloader.spec.ts:
```diff
@@ -196,6 +200,272 @@ describe('RouterPreloader', () => {
+    const lazyLoadChildrenSpy = jasmine.createSpy('lazymodule');
+    const buildPreloader = (
+        delayObserver$: Observable<string[]>,
+        ): typeof PreloadingStrategy => {
```
```suggestion
        ): PreloadingStrategy => {
```

I think this might be easier to understand if you use `useFactory` instead and have the builder just return an instance.

atscott commented on packages/router/test/router_preloader.spec.ts:
```diff
@@ -196,6 +200,272 @@ describe('RouterPreloader', () => {
+    });
+
+    it('without reloading loaded modules',
+       fakeAsync(inject([RouterPreloader, Router], (preloader: RouterPreloader, router: Router) => {
```
```suggestion
       fakeAsync(() => {
       const preloader = TestBed.inject(RouterPreloader);
       const router = TestBed.inject(Router);
```
let's convert these to the more common `TestBed.inject` rather than the old-school `inject([]`) notation

atscott commented on packages/router/test/router_preloader.spec.ts:
```diff
@@ -196,6 +200,272 @@ describe('RouterPreloader', () => {
            })));
   });
 
+  describe('should support preloading stratages', () => {
```
```suggestion
  describe('should support preloading strategies', () => {
```

atscott commented on packages/router/test/router_preloader.spec.ts:
```diff
@@ -196,6 +200,272 @@ describe('RouterPreloader', () => {
+                return fn().pipe(
+                    catchError(() => of(null)),
+                );
+              }));
```
```suggestion
              }),);
```
nit: adding a trailing comma will put each of the operators on a new line when formatted and it'll be a bit easier to read.

atscott commented on packages/router/test/router_preloader.spec.ts:
```diff
@@ -196,6 +200,272 @@ describe('RouterPreloader', () => {
+       })));
+
+
+    it('and cope with the loader throwing exceptions from prerouter but allow retry',
```
```suggestion
    it('and cope with the loader throwing exceptions from preloader but allow retry',
```

pgammans replied:
I changed this again to  'and cope with the loader throwing exceptions during module load but allow retry'
as this checks that errors caught from the loadChildren call/promise on the route.
ATM we dont chek we support handling error from within the preloader. Do you want another test with a preloader the caused an exception.

atscott replied:
What you have is fine. I thought it was a typo.

atscott commented on packages/router/test/router_preloader.spec.ts:
```diff
@@ -196,6 +200,272 @@ describe('RouterPreloader', () => {
+
+         lazyLoadChildrenSpy.and.returnValue(of(LoadedModule1));
+         router.navigateByUrl('/lazy/LoadedModule1').catch(() => {
+           expect('Not to get here').toBeUndefined('Not to throw');
```
```suggestion
           fail('navigation should not throw');
```

atscott commented on packages/router/test/router_preloader.spec.ts:
```diff
@@ -196,6 +200,272 @@ describe('RouterPreloader', () => {
+         expect(lazyLoadChildrenSpy).toHaveBeenCalledTimes(1);
+         lazyLoadChildrenSpy.and.returnValue(of(LoadedModule1));
+         router.navigateByUrl('/lazy/LoadedModule1').catch(() => {
+           expect('Not to get here').toBeUndefined('Not to throw');
```
```suggestion
           fail('navigation should not throw');
```

atscott commented on packages/router/test/router_preloader.spec.ts:
```diff
@@ -196,6 +200,272 @@ describe('RouterPreloader', () => {
+           'RouteConfigLoadStart(path: lazy)', 'RouteConfigLoadEnd(path: lazy)'
+         ]);
+
+         // Release submodule to check it dose infact load
```
```suggestion
         // Release submodule to check it does in fact load
```

atscott reviewed (changes requested):
@pgammans - Thanks for resubmitting this PR with the updates! A few small nits in the test and one question about finding a strategy to avoid using `share` but otherwise LGTM

atscott commented on packages/router/src/router_config_loader.ts:
```diff
@@ -7,8 +7,8 @@
-import {from, Observable, of} from 'rxjs';
-import {map, mergeMap} from 'rxjs/operators';
+import {from, Observable, of, Subject} from 'rxjs';
+import {catchError, map, mergeMap, share, tap} from 'rxjs/operators';
```
```suggestion
import {catchError, map, mergeMap, tap} from 'rxjs/operators';
```

atscott commented:
@pgammans one more thing - could you squash all of your commits into one?

atscott commented:
[global presubmit](https://test.corp.google.com/ui#id=OCL:351668559:BASE:351784848:1610664049040:7f031cd8)

atscott commented:
Hi @pgammans - sorry for the slow response. After running internal tests, the approach we made to avoid introducing the `share` operator caused failures. Using the `share` like was there initially _does_ work though. 

I don't know if we can justify the additional payload size increase to address this edge case where a preloader and a navigation duplicate the loading of a lazy route. That would increase the payload for everyone and only prevent an extra load of a lazy module in a rare scenario so I think the cost-benefit doesn't quite work out in our favor.

What we can certainly do now though is keep the check that you added in `preloadConfig` to partially address the issue.

pgammans commented:
@atscott can i ask what the failure was and to see if i can develop a third solution that addresses all the issues.

Obviously share does something that addresses this issue it may be possible to implement a fix for this without all the overheads that the generic share function will need to address and 
 ```
 load(parentInjector: Injector, route: Route): Observable<LoadedRouterConfig> {
    if (route._loader$) {
      return route._loader$;
    }

    if (this.onLoadStartListener) {
      this.onLoadStartListener(route);
    }
    const moduleFactory$ = this.loadModuleFactory(route.loadChildren!);
    route._loader$ =  new ConnectableObservable(
      moduleFactory$.pipe(
        map((factory: NgModuleFactory<any>) => {
          if (this.onLoadEndListener) {
            this.onLoadEndListener(route);
          }
          const module = factory.create(parentInjector);
          // When loading a module that doesn't provide `RouterModule.forChild()` preloader will get
          // stuck in an infinite loop. The child module's Injector will look to its parent
          // `Injector` when it doesn't find any ROUTES so it will return routes for it's parent
          // module instead.or
          return new LoadedRouterConfig(
              flatten(
                  module.injector.get(ROUTES, undefined, InjectFlags.Self | InjectFlags.Optional))
                  .map(standardizeConfig),
              module);
        }),
        catchError((err) => {
          route._loader$ = undefined;
          throw err;
        }),
      ),
      () => new Subject<LoadedRouterConfig>()
    ).pipe(refCount())
    return route._loader$;

```
Even this may be smaller as remove some of share features.
How can i test the size change?

atscott commented:
@pgammans - I haven't been able to debug the exact issue internally so I unfortunately can't provide much more information now. It's a few of the application's webdriver integration test that are failing and I'll have to figure out how to spin up the app locally to debug it. 
The snippet with `ConnectableObservable` and `refCount` also passes the tests but I'm still concerned about the tradeoff in bundle size vs impact. Running the bundle size check locally doesn't seem to work (on our plate to investigate soon) but if you make the change to your PR, the CI will be able to run the test and tell you the new bundle size.

pgammans commented:
@atscott
I assume this is OK, CI says no size change. I assume this is as `ConnectableObservable` is already allocated in the routers golden file.

atscott commented on packages/router/src/router_config_loader.ts:
```diff
@@ -28,27 +28,41 @@ export class RouterConfigLoader {
+    // Use custom ConnectableObservable as share increasing the bundle size too much
+    route._loader$ =
+        new ConnectableObservable(
+            moduleFactory$.pipe(
```
suggestion: What do you think of extracting this first argument of the `ConnectableObservable` to a variable so it's easier to see the two arguments?

pgammans replied:
I've push a change for this and rebased to master.
I'm not 100% sure i agree as i think it breaks the link the the effective 'share' is part of the loader, but i can see an argument for separation of the loader from the share.

atscott commented:
@pgammans - Yep, I believe you are correct.  It looks like the goldens also already have the refcount operator. I want to run some final tests on this but at the moment it's looking good. Thanks for the work on this!

atscott commented:
[global presubmit](https://test.corp.google.com/ui#id=OCL:351668559:BASE:355057848:1612227711256:ede2ab05) (all tests passing)

atscott commented on packages/router/src/router_config_loader.ts:
```diff
@@ -28,27 +28,39 @@ export class RouterConfigLoader {
       private onLoadEndListener?: (r: Route) => void) {}
 
   load(parentInjector: Injector, route: Route): Observable<LoadedRouterConfig> {
+    if (route._loader$) {
```
Now that this is no longer a `replay`, does it make sense to return the `route._loadedConfig` instead if it exists or check if the `route._loader$` is already complete? The callers of `load` here prevent this from being the case, but it might still be a good safeguard. What do you think?

pgammans replied:
Yes we need to return `route._loadedConfig` but as `route._loadedConfig` will only exist once `route._loader$ `is complete.
Currently it is the callers of load need to set `route._loadedConfig` themselves. which is a bit odd and and addressed in #36654

The callers are (possibly more)
- `RouterPreloader.preloadConfig` 
- the pipeline of the routeGuard from `ApplyRedirects.getChildConfig`
- the pipeline of the `ApplyRedirects.matchSegmentAgainstRoute`

The first two have checks for `route._loadedConfig` and use that if available so load will not get called in these cases. The later is probably should be updated to  but atm the check in load would prevent duplicate loads

  ```
  private matchSegmentAgainstRoute(
        ngModule: NgModuleRef<any>, rawSegmentGroup: UrlSegmentGroup, route: Route,
        segments: UrlSegment[], outlet: string): Observable<UrlSegmentGroup> {
      if (route.path === '**') {
        if (route.loadChildren) {
+         const loaded$ = route._loadedConfig ? of(route._loadedConfig) :
                                            this.configLoader.load(ngModule.injector, route);
-         return this.configLoader.load(ngModule.injector, route)
-              .pipe(map((cfg: LoadedRouterConfig) => {
+        return loaded.pipe(map((cfg: LoadedRouterConfig) => {
                route._loadedConfig = cfg;
                return new UrlSegmentGroup(segments, {});
              }));
        }

        return of(new UrlSegmentGroup(segments, {}));
      }
      ...
```

Lastly in the delay between create and compete of `route._loader$` we need to use the same observable if load is called multiple times. Which it can be as, in fact we have test that do this.
ie there can be a call from multiple of the load callers. In the test we have one from the call `RouterPreloader.preloadConfig`  and on from the call through `ApplyRedirects.getChildConfig`

atscott replied:
I absolutely agree that it's odd for the callers of `load` to set `_loadedConfig` rather than having `RouteConfigLoader`  and I like the direction of #36654.

Since `RouterConfigLoader` isn't public API and neither are the callers of `load`, I'll leave it to you to decide which approach to take to ensure we aren't returning the already completed `_loader` observable:
* Handle it in the `load` function in the same manner the callers are doing by checking for `_loadedConfig`
* Update `ApplyRedirects` to not call `load` if `_loadedConfig` already exists.

We can then consolidate everything in #36654 or in another PR that omits the `RouteConfigReady` feature and only improves the DRY-ness (would be easier to land, as it's not a public api change).

pgammans replied:
@atscott Ideally I think we should put the test `route._loadedConfig` inside load, and move it setting to part of the load, but `apply_redirects` use the status of  `route._loadedConfig` to avoid duplicating calls to the load guard.

Thus we'd need some another call to allow us prevent calling the guard twice. ie would probably need to be change to public API of Route!

So i propose: (basically what we have.)
* load : Allows overlapping calls and will only load once.
* load : decide what to do if the previous load is closed / finished and load called
   - throw
   - reload the module
   - just return the current loaded module, but dose this also sent event etc again?
* leave it up to the callers of load to manage checking of `route._loadedConfig` 
 
Once landed look at improving setting of `route._loadedConfig` to be part of the loads. observable pipe (new PR based on #36654)

I added a check to `ApplyRedirects.matchSegmentAgainstRoute` and a test using this code path.

atscott replied:
@pgammans - 👍  Agreed.

I think maybe you forgot to push up your changes to `ApplyRedirects.matchSegmentAgainstRoute`?  After that's done, I'll run a final sanity check suite of internal tests and this should be good to go! Thanks for the hard work!

atscott commented:
[presubmit](https://test.corp.google.com/ui#id=OCL:355874449:BASE:355874472:1612548565934:f569c753)

pgammans commented:
@atscott Is there anything else I need to do on this, or is it just a case of waiting for the backlog to reduce?

I also see you pass the odd failure of ci/circleci: test (ci seams to random fail some tests) i assume i don't need to re-push as i cannot cause ci to retest.

As once this is merged I will pull out the `route._loadedConfig` changes we discussed and prepare a PR for that.

atscott commented:
@pgammans The failing test is a size regression test so we're back to the problem of this change being too big for the issue it fixes. My guess is that initially, the change was made right after the sizes were updated and it is just barely under the 500KB threshold. The current size at `head` is 241330 and the size with this change is 241843. You can get these numbers by running `yarn bazel run //integration:cli-hello-world-lazy_test.debug` (I've been able to get this to work on Linux, but not my Mac).

I don't think this change can be merged because it's a ~500KB increase  (uncompressed, but still an additional 137KB compressed) for all applications that use the router and the only benefit it provides is that it prevents overlapping loads in the rare case that this might happen.

atscott commented:
Ah, my mistake! I was reading this wrong and it's really just 500 _bytes_! This is pretty much expected and unavoidable. I also tested this with the `shareReplay` and also as expected, the increase was _much_ more: 247007, closer to 5000 bytes. I'll push a fixup commit shortly to update the value in size tracking. Then this will go through a size change approval.

atscott reviewed (approved):
reviewed-for: fw-router
reviewed-for: size-tracking

Note to size tracking reviewers: the size increase is 513 bytes, uncompressed. This seems reasonable enough to me and unavoidable in the end in order to coordinate between duplicate module loads.

IgorMinar reviewed (approved):
LGTM 

Reviewed-for: size-tracking

petebacondarwin commented on packages/router/test/apply_redirects.spec.ts:
```diff
@@ -514,6 +514,45 @@ describe('applyRedirects', () => {
          expect(loaded).toEqual(['root', 'aux']);
        }));
 
+    it('should not try to load any matching configuration if previous load competed',
```
```suggestion
    it('should not try to load any matching configuration if previous load completed',
```

petebacondarwin reviewed (approved):
Reviewed-for: size-tracking

pgammans commented:
@petebacondarwin  I have applied your suggested commit for the spelling mistake, rebased onto master. I have also squashed all the commits as I assume this would be wanted.

pgammans commented:
@atscott  the size of aio-local-viewengine has also increased do you want a commit to increase ti size too?

expected 432078, actual: 432647 so an increase of 569 bytes

atscott committed:
fixup! fix(router): fix load interaction of navigation and preload strategies

atscott commented:
FYI - this fix was reverted as there were internal failures during the sync that need further investigation.

atscott commented:
Okay, so the issue is that by de-duplicating the module loading and using the same router config objects, we change how routes are re-used. That is, a route that was not previously re-used will now be re-used because the route objects have the same reference. 
This is actually the correct behavior, but it's something we'll still have to address. Devs may need to update their tests if they accidentally rely on the current broken behavior. I'm still investigating how widespread this problem is.

atscott commented:
Update: There was only one test failure (the one from yesterday) in the global test results. That indicates the issue isn't widespread enough to require further action other than a fix to this one test. After that is done, we can resubmit this PR.

atscott commented:
[global presubmit](https://test.corp.google.com/ui#id=OCL:357091291:BASE:357736336:1613496667740:a0693fef)

atscott commented:
caretaker note: Since this PR previously had to be rolled back, I think it'd be a good idea to merge and sync on its own.
